tee-worker/omni-executor/evm-aggregators-client/src/evm_tx_manager.rs

```diff
@@ -0,0 +1,459 @@
+	OkxClient: OkxSwap + ?Sized,
+{
+	pub async fn get_gas_price_by_level(&self, chain_id: u64, level: i32) -> Result<Decimal, ()> {
+		let gas_price = self
+			.okx_client
+			.get_gas_price(chain_id)
+			.await
+			.map_err(|e| error!("Failed to get price by level due to: {:?}", e))?;
+
+		match level {
+			1_i32 => Decimal::from_str(&gas_price.min)
+				.map_err(|_| error!("Failed to get price by level")),
+			2_i32 => Decimal::from_str(&gas_price.normal)
+				.map_err(|_| error!("Failed to get price by level")),
+			3_i32 => Decimal::from_str(&gas_price.max)
+				.map_err(|_| error!("Failed to get price by level")),
+			_ => {
+				error!("Level {} not found", level);
+				Err(())
+			},
+		}
+	}
+}
+
+impl<EthClient, InchClient, KyberClient, OkxClient>
+	EvmTxManager<EthClient, InchClient, KyberClient, OkxClient>
+where
+	EthClient: RpcProvider<Addr = Address> + ?Sized + EthereumClient,
+	InchClient: InchSwap + ?Sized,
+	KyberClient: KyberSwap + ?Sized,
+	OkxClient: OkxSwap + ?Sized,
+{
+	#[allow(dead_code)]
+	async fn construct_unsigned_market_tx(
+		&self,
+		tx: CreateMarketTx,
+	) -> Result<Vec<TransactionRequest>, ()> {
+		let amount_decimal = Decimal::from_str(&tx.amount_in).unwrap();
+		let multiplier = DECIMALS_TO_VALUE.get(&tx.in_decimal).cloned().unwrap_or(1);
+		let amount_decimal = amount_decimal * Decimal::from(multiplier);
+
+		let from = Address::from_hex(&tx.user_wallet_address)
+			.map_err(|e| error!("Failed to convert user wallet to address: {}", e))?;
+
+		let mut balance = self.eth_client.get_balance(from).await.map_err(|_| {
+			error!("Couldn't get balance");
+		})?;
+
+		let mut nonce = self.eth_client.get_pending_nonce(from).await.map_err(|_| {
+			error!("Couldn't get pending nonce");
+		})?;
+
+		let is_buy = is_native_token(&tx.in_token_ca.clone().into_bytes());
+		let platform = Platform::KyberSwap;
+
+		let amount_u128 = amount_decimal.to_u128().ok_or_else(|| {
+			error!("amount_decimal could not be converted to u128: {}", amount_decimal);
+		})?;
+
+		let amount = Uint::try_from(amount_u128).map_err(|e| {
+			error!("Failed to convert amount_u128 to Uint: {:?}", e);
+		})?;
+
+		let mut transactions: Vec<TransactionRequest> = Vec::new();
+		if is_buy {
+			if balance < amount {
+				error!("Balance is too low");
+				return Err(());
+			}
+			balance -= amount;
+		} else {
+			let approve_addr: Address = match platform {
+				Platform::KyberSwap => {
+					Address::from_hex(KYBER_SWAP_APPROVE_ADDRESS).map_err(|_| {
+						error!("Failed to convert address from hex to address");
+					})?
+				},
+				Platform::Inch => Address::from_hex(INCH_SWAP_APPROVE_ADDRESS).map_err(|_| {
+					error!("Failed to convert address from hex to address");
+				})?,
+				Platform::Okx => Address::from_hex(OKX_SWAP_APPROVE_ADDRESS).map_err(|_| {
+					error!("Failed to convert address from hex to address");
+				})?,
+			};
+
+			let approve_tx = self
+				.eth_client
+				.construct_approve_erc20_tx(
```

Comment:
Is this approval mandatory? It seems to be executed for every swap, but I don't recall we have such flow in the current impl

Reply:
We don't need this for every swap, if it's `BUY` we just check balance, We need this for `SELL` , that is we approve ERC20 on behalf of the aggregators approve contract, so that we can then proceed with a swap transaction. I think that's the general flow of using a swap in EVM chains, we first approve and then swap.